filament/src/Tonemapper.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "details/Tonemapper.h"
+
+#include "details/Engine.h"
+
+#include <image/ColorTransform.h>
+
+#include <utils/JobSystem.h>
+
+// When defined, the ACES tone mapper will match the brightness of the "ACES sRGB" tone mapper
+// It is *not* correct, but it helps for compatibility
+#define TONEMAP_ACES_MATCH_BRIGHTNESS
+
+#include "ACES.h"
+
+namespace filament {
+
+using namespace utils;
+using namespace math;
+using namespace backend;
+
+static constexpr size_t LUT_DIMENSION = 32u;
+
+Tonemapper::Tonemapper(FEngine& engine) {
+    DriverApi& driver = engine.getDriverApi();
+
+    // it's okay to allocate the data in the command stream because it's only 256 KiB
```

Comment:
What does it mean to "allocate the data in the command stream"?

Reply:
how I should have removed this comment, I'm using malloc now.

you can call driver.allocate() to allocate memory in the command buffer, which has the same lifetime as the command buffer itself (so you don't need a callback to free it).